File: api/tests/fixtures/graph_response.py
import datetime
import uuid

from database.pg.graph_ops.graph_crud import CompleteGraph
from database.pg.models import Edge, Graph, Node

GRAPH_ID = uuid.UUID("11111111-2222-3333-4444-555555555555")
USER_ID = uuid.UUID("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee")
FOLDER_ID = uuid.UUID("12345678-1234-5678-1234-567812345678")
WORKSPACE_ID = uuid.UUID("87654321-4321-8765-4321-876543218765")


def deterministic_reply(size: int, seed: int = 0) -> str:
    prefix = "The graph response preserves Unicode café 🚀 and code exactly.\n"
    remaining_bytes = size - len(prefix.encode("utf-8"))
    if remaining_bytes < 0:
        raise ValueError("reply size is too small for the deterministic prefix")

    parts = [prefix]
    counter = seed * 10_007
    while remaining_bytes:
        checksum = (counter * 2_654_435_761) & 0xFFFFFFFF
        line = (
            f"Segment {counter:09d}: analyze request_{counter % 997:03d}; "
            f"value=0x{checksum:08x}; result={(checksum ^ seed) % 100_003:05d}.\n"
        )
        chunk = line[:remaining_bytes]
        parts.append(chunk)
        remaining_bytes -= len(chunk)
        counter += 1
    return "".join(parts)


def build_graph_fixture(
    *, node_count: int, edge_count: int, reply_count: int, reply_size: int
) -> CompleteGraph:
    graph = Graph(
        id=GRAPH_ID,
        user_id=USER_ID,
        folder_id=FOLDER_ID,
        workspace_id=WORKSPACE_ID,
        name="Deterministic graph",
        description=None,
        temporary=False,
        pinned=False,
        created_at=datetime.datetime(2026, 7, 15, 12, 30, tzinfo=datetime.timezone.utc),
        updated_at=datetime.datetime(2026, 7, 15, 12, 45, tzinfo=datetime.timezone.utc),
        custom_instructions=[],
        max_tokens=None,
        temperature=0.25,
        top_p=None,
        top_k=40,
        frequency_penalty=None,
        presence_penalty=0.0,
        repetition_penalty=None,
        reasoning_effort="high",
    )
    graph.node_count = node_count

    nodes = []
    for index in range(node_count):
        data: dict[str, object] = {
            "title": f"Node {index}",
            "nested": {
                "null": None,
                "false": False,
                "zero": 0,
                "empty_string": "",
                "empty_object": {},
                "empty_array": [],
            },
        }
        if index < reply_count:
            data["reply"] = deterministic_reply(reply_size, seed=index)
        nodes.append(
            Node(
                id=f"node-{index}",
                graph_id=GRAPH_ID,
                type="chat" if index % 3 else "prompt",
                position_x=float(index * 17),
                position_y=float(index * -9),
                width="320px" if index % 11 == 0 else "100px",
                height="180px" if index % 13 == 0 else "100px",
                parent_node_id=f"node-{index - 1}" if index and index % 7 == 0 else None,
                data=data,
            )
        )

    edges = []
    for index in range(edge_count):
        source_index = index % node_count
        target_index = (source_index + 1) % node_count
        edges.append(
            Edge(
                id=f"edge-{index}",
                graph_id=GRAPH_ID,
                source_node_id=f"node-{source_index}",
                target_node_id=f"node-{target_index}",
                source_handle_id=f"source-{index}" if index % 5 == 0 else None,
                target_handle_id=f"target-{index}" if index % 7 == 0 else None,
                type="smoothstep" if index % 4 == 0 else None,
                label=f"Edge {index}" if index % 9 == 0 else None,
                animated=index % 6 == 0,
                style={"strokeWidth": 0, "visible": False} if index % 8 == 0 else None,
                data={"empty": "", "items": []} if index % 10 == 0 else None,
                markerEnd={"type": "arrowclosed"},
            )
        )

    return CompleteGraph(graph=graph, nodes=nodes, edges=edges)


def build_small_graph_fixture() -> CompleteGraph:
    return build_graph_fixture(node_count=30, edge_count=38, reply_count=4, reply_size=4096)


def build_large_graph_fixture() -> CompleteGraph:
    return build_graph_fixture(node_count=500, edge_count=650, reply_count=220, reply_size=20_000)

File: ui/app/utils/graphResponse.ts
import { REASONING_EFFORTS } from '@/types/enums';
import type { ReasoningEffortEnum } from '@/types/enums';
import {
    GRAPH_EDITOR_RESPONSE_VERSION,
    type DecodedGraphEditorEdgeV1,
    type DecodedGraphEditorNodeV1,
    type DecodedGraphEditorResponse,
    type GraphJsonContainer,
} from '@/types/graphResponse';

const REASONING_EFFORT_VALUES = new Set<string>(REASONING_EFFORTS);

const isRecord = (value: unknown): value is Record<string, unknown> =>
    typeof value === 'object' && value !== null && !Array.isArray(value);

const invalid = (path: string, expected: string): never => {
    throw new Error(`Invalid graph response value at ${path}: expected ${expected}`);
};

const requiredString = (value: unknown, path: string): string => {
    if (typeof value !== 'string') return invalid(path, 'a string');
    return value;
};

const requiredNumber = (value: unknown, path: string): number => {
    if (typeof value !== 'number' || !Number.isFinite(value)) {
        return invalid(path, 'a finite number');
    }
    return value;
};

const requiredInteger = (value: unknown, path: string): number => {
    const number = requiredNumber(value, path);
    if (!Number.isSafeInteger(number)) return invalid(path, 'a safe integer');
    return number;
};

const optionalNullableString = (value: unknown, path: string): string | null => {
    if (value === undefined || value === null) return null;
    return requiredString(value, path);
};

const optionalNullableNumber = (value: unknown, path: string): number | null => {
    if (value === undefined || value === null) return null;
    return requiredNumber(value, path);
};

const optionalNullableInteger = (value: unknown, path: string): number | null => {
    if (value === undefined || value === null) return null;
    return requiredInteger(value, path);
};

const optionalBoolean = (value: unknown, path: string, fallback: boolean): boolean => {
    if (value === undefined) return fallback;
    if (typeof value !== 'boolean') return invalid(path, 'a boolean');
    return value;
};

const optionalJsonContainer = (value: unknown, path: string): GraphJsonContainer | null => {
    if (value === undefined || value === null) return null;
    if (Array.isArray(value) || isRecord(value)) return value;
    return invalid(path, 'an object, array, or null');
};

const optionalReasoningEffort = (value: unknown, path: string): ReasoningEffortEnum | null => {
    if (value === undefined || value === null) return null;
    if (typeof value !== 'string' || !REASONING_EFFORT_VALUES.has(value)) {
        return invalid(path, 'a supported reasoning effort or null');
    }
    return value as ReasoningEffortEnum;
};

const customInstructions = (value: unknown, path: string): string[] => {
    if (value === undefined) return [];
    if (!Array.isArray(value)) return invalid(path, 'an array of strings');
    return value.map((instruction, index) =>
        requiredString(instruction, `${path}[${index}]`),
    );
};

const decodeNode = (value: unknown, index: number): DecodedGraphEditorNodeV1 => {
    const path = `nodes[${index}]`;
    if (!isRecord(value)) return invalid(path, 'an object');

    return {
        id: requiredString(value.id, `${path}.id`),
        type: requiredString(value.type, `${path}.type`),
        position_x: requiredNumber(value.position_x, `${path}.position_x`),
        position_y: requiredNumber(value.position_y, `${path}.position_y`),
        width:
            value.width === undefined
                ? '100px'
                : requiredString(value.width, `${path}.width`),
        height:
            value.height === undefined
                ? '100px'
                : requiredString(value.height, `${path}.height`),
        parent_node_id: optionalNullableString(value.parent_node_id, `${path}.parent_node_id`),
        data: optionalJsonContainer(value.data, `${path}.data`),
    };
};

const decodeEdge = (value: unknown, index: number): DecodedGraphEditorEdgeV1 => {
    const path = `edges[${index}]`;
    if (!isRecord(value)) return invalid(path, 'an object');

    return {
        id: requiredString(value.id, `${path}.id`),
        source_node_id: requiredString(value.source_node_id, `${path}.source_node_id`),
        target_node_id: requiredString(value.target_node_id, `${path}.target_node_id`),
        source_handle_id: optionalNullableString(
            value.source_handle_id,
            `${path}.source_handle_id`,
        ),
        target_handle_id: optionalNullableString(
            value.target_handle_id,
            `${path}.target_handle_id`,
        ),
        type: optionalNullableString(value.type, `${path}.type`),
        label: optionalNullableString(value.label, `${path}.label`),
        animated: optionalBoolean(value.animated, `${path}.animated`, false),
        style: optionalJsonContainer(value.style, `${path}.style`),
        data: optionalJsonContainer(value.data, `${path}.data`),
    };
};

export const decodeGraphEditorResponse = (value: unknown): DecodedGraphEditorResponse => {
    if (!isRecord(value)) return invalid('response', 'an object');
    if (value.version !== GRAPH_EDITOR_RESPONSE_VERSION) {
        throw new Error(`Unsupported graph response version: ${String(value.version)}`);
    }
    if (!isRecord(value.graph)) return invalid('graph', 'an object');
    if (!Array.isArray(value.nodes)) return invalid('nodes', 'an array');
    if (!Array.isArray(value.edges)) return invalid('edges', 'an array');

    const graph = value.graph;
    return {
        version: GRAPH_EDITOR_RESPONSE_VERSION,
        graph: {
            id: requiredString(graph.id, 'graph.id'),
            name: requiredString(graph.name, 'graph.name'),
            node_count: requiredInteger(graph.node_count, 'graph.node_count'),
            folder_id: optionalNullableString(graph.folder_id, 'graph.folder_id'),
            workspace_id: optionalNullableString(graph.workspace_id, 'graph.workspace_id'),
            description: optionalNullableString(graph.description, 'graph.description'),
            temporary: optionalBoolean(graph.temporary, 'graph.temporary', false),
            pinned: optionalBoolean(graph.pinned, 'graph.pinned', false),
            created_at: optionalNullableString(graph.created_at, 'graph.created_at'),
            updated_at: optionalNullableString(graph.updated_at, 'graph.updated_at'),
            custom_instructions: customInstructions(
                graph.custom_instructions,
                'graph.custom_instructions',
            ),
            max_tokens: optionalNullableInteger(graph.max_tokens, 'graph.max_tokens'),
            temperature: optionalNullableNumber(graph.temperature, 'graph.temperature'),
            top_p: optionalNullableNumber(graph.top_p, 'graph.top_p'),
            top_k: optionalNullableInteger(graph.top_k, 'graph.top_k'),
            frequency_penalty: optionalNullableNumber(
                graph.frequency_penalty,
                'graph.frequency_penalty',
            ),
            presence_penalty: optionalNullableNumber(
                graph.presence_penalty,
                'graph.presence_penalty',
            ),
            repetition_penalty: optionalNullableNumber(
                graph.repetition_penalty,
                'graph.repetition_penalty',
            ),
            reasoning_effort: optionalReasoningEffort(
                graph.reasoning_effort,
                'graph.reasoning_effort',
            ),
        },
        nodes: value.nodes.map(decodeNode),
        edges: value.edges.map(decodeEdge),
    };
};
